### PR TITLE
Closes #654: On presave, check if there is an overridden thumb, if there is replace it on the video

### LINKED ISSUE #654: Add ability to override remote video thumbnail with locally added thumnail.
## Motivation
When youtube sends a video thumbnail often it is not cropped to reasonable dimensions for how we are using it in text over media background.

## Proposed Resolution
Add a way to upload a new image over the default oembed thumb. 

### Describe the solution you'd like
Should probably implement something like this to override the cover image: https://drupal.stackexchange.com/questions/286291/custom-remote-video-thumbnail

### Roles and Permissions considerations
A clear and concise description of how each of the following roles would be impacted by this change:
- Anonymous user - view new image
- Authenticated user- view new image
- Content editor - upload and crop new image
- Content administrator - upload and crop new image
- Administrator - upload and crop new image

### PR DESCRIPTION
**Pull request by trackleft**
## Description
Closes https://github.com/az-digital/az_quickstart/issues/654
## Related Issue

Depends on #1059 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by trackleft** [comment]
This is ready for review now https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-913.probo.build/user/login?destination=/admin/content/media

**Comment by trackleft** [comment]
Probo tests pass https://app.probo.ci/#/dashboard/build/396d3a87-8c33-454b-b411-d8409bffa699

**Comment by smayden** [comment]
I've tried testing this and had a difficult time figuring out how to to add the thumbnail (I had to go in and do it through files and edit the media).

But even when doing that the thumbnail I added did not appear on the video (even after cache clear).

**Comment by trackleft** [comment]
> I've tried testing this and had a difficult time figuring out how to to add the thumbnail (I had to go in and do it through files and edit the media).
> 
> But even when doing that the thumbnail I added did not appear on the video (even after cache clear).

All of what you are saying is true, I opted not to add the media file to the library view mode because it does almost nothing at the moment, but it does do the thing that it is supposed to do, which is replace the thumbnail.

In a future ticket for text with media, the plan is to utilize this functionality. We can leave it out for now if you want.

File: modules/custom/az_media/config/install/field.field.media.az_remote_video.field_media_az_image.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+bundle: az_remote_video
+label: 'Thumbnail or cover image'
+description: 'Optionally, override the default thumbnail image from your video hosting provider.'
+required: false
+translatable: true
+default_value: {  }
+default_value_callback: ''
+settings:
+  file_directory: '[date:custom:Y]-[date:custom:m]'
+  file_extensions: 'png gif jpg jpeg webm'
```
Comment (trackleft) [inline modules/custom/az_media/config/install/field.field.media.az_remote_video.field_media_az_image.yml:21]:
```suggestion
  file_extensions: 'png gif jpg jpeg webp'
```

**Review by smayden — APPROVED** [review]
tested that thumbnail changes on media item (not on text on media yet)

File: modules/custom/az_media/az_media.module [hunk 2]
```diff
@@ -148,3 +149,18 @@ function az_media_theme() {
...
+ * Implements hook_entity_presave().
+ *
+ * This allows content editors to override remote video thumbs.
+ */
+function az_media_entity_presave(EntityInterface $entity) {
+  if (
+    $entity->getEntityTypeId() === 'media'
+    && $entity->bundle() === 'az_remote_video'
+    && isset($entity->field_media_az_image->entity)
+    ) {
```
Comment (joeparsons) [inline modules/custom/az_media/az_media.module:163]: Nitpick:  Line breaks like these in control structures are discouraged in the coding standards:
https://www.drupal.org/docs/develop/standards/coding-standards#linelength

```suggestion
  if ($entity->getEntityTypeId() === 'media' && $entity->bundle() === 'az_remote_video' && isset($entity>field_media_az_image->entity)) {
```

File: modules/custom/az_media/az_media.module [hunk 3]
```diff
@@ -148,3 +149,14 @@ function az_media_theme() {
...
   ];
 }
+
+/**
+ * Implements hook_entity_presave().
+ *
+ * This allows content editors to override remote video thumbs.
+ */
+function az_media_entity_presave(EntityInterface $entity) {
+  if ($entity->getEntityTypeId() === 'media' && $entity->bundle() === 'az_remote_video' && isset($entity>field_media_az_image->entity)) {
```
Comment (joeparsons) [inline modules/custom/az_media/az_media.module:159]: Oops, my suggestion messed this up.
```suggestion
  if ($entity->getEntityTypeId() === 'media' && $entity->bundle() === 'az_remote_video' && isset($entity->field_media_az_image->entity)) {
```

Comment (joeparsons) [reply modules/custom/az_media/az_media.module:159]: (PHPCS failing on this now)

**Review by smayden — DISMISSED** [review]
I was able to pull this into a local environment along with the companion PR #1059. Verified that I can now switch out the thumbnail and it displays on text on media for desktop/mobile (with no more black bars on mobile).

Before
image

After with these two PRs
image

**Review by joeparsons — APPROVED** [review]
Looks good to me now (code-wise).

**Review by smayden — APPROVED** [review]

**Review by mmunro-ltrr — APPROVED** [review]

**Pull request merged**